### ISSUE TITLE
Formal specification of automatic type coercions

### ISSUE THREAD

**Issue by jdidion** [openwdl/wdl#396]
This proposal is to add details to the specification on what type coercions are applied automatically, and, where necessary, to specify the order of precedence of coercions when multiple valid coercions are possible. I am listing the coercions here, but in the spec they should be presented as a table.

# General rules

* Integer may be coerced to Float.
* Float may not be coerced to Int.
* String may be coerced to File.
* String may be coerced to Directory.
* String may not be coerced to numeric types.
* Any primitive type may be coerced to String.
* There are no types that automatically coerce to Boolean.
* `X?` to `Y?` is allowed in the case where `X` is coercible to `Y`.
* `X` may be coerced to `X?`.
* `X?` may not be coerced to `X`.
* `Array[X]` to `Array[Y]` is allowed in the case where `X` is coercible to `Y`.
* `Array[X]+` to `Array[X]` is allowed.
* `Map[W, X]` to `Map[Y, Z]` is allowed in the case where `W` is coercible to `Y` and `X` is coercible to `Z`.
* `Pair[W, X]` to `Pair[Y, Z]` is allowed in the case where `W` is coercible to `Y` and `X` is coercible to `Z`.
* Object may be coerced to a user-defined data type so long as 1. the object's keys are a subset of the struct's members, 2. any keys missing in the object are declared as optional in the struct, and 3. every value in the object Is coercible to the data type of the corresponding member in the struct.
* No other coercions are allowed between complex types.

# Limited exceptions

Implementers may choose to allow limited exceptions to the above rules, with the understanding that workflows depending on these exceptions may not be portable. These exceptions are provided for backward-compatibility, are considered deprecated, and will be removed in a future version of WDL.

*  Float to Int, when the coercion can be performed with no loss of precision, e.g. `1.0 -> 1`.
* String to Int/Float, when the coercion can be performed with no loss of precision.
* `X?` may be coerced to `X`, and an error is raised if the value is undefined.
* `Array[X]` to `Array[X]+`, when the array is non-empty (an error is raised otherwise).
* `Map[W, X]` to `Array[Pair[Y, Z]]`, in the case where `W` is coercible to `Y` and `X` is coercible to `Z`.
* `Array[Pair[W, X]]` to `Map[Y, Z]`, in the case where `W` is coercible to `Y` and `X` is coercible to `Z`.
* Map to Object, in the case of `Map[String, X]`.
* Map to struct, in the case of `Map[String, X]` where all members of the struct have type `X`.
* Object to `Map[String, X]`, in the case where all object values are of the same type.

# Order of precedence

There are some operators for which it is possible to coerce the same arguments in multiple different ways. For such operators, it is necessary to define the order of preference so that a single function prototype can be selected from among the available options for any given set of arguments.

The `+` operator is overloaded for both numeric addition and String concatenation. This can lead to the following kinds of situations:

```
String s = "1.0"
Float f = 2.0
String x = "${s + f}"
```

There are two possible ways to evaluate the `s + f` expression:

1. Coerce `s` to `Float` and perform floating point addition, then coerce to `String` with the result being `x = "3.0"`.
2. Coerce `f` to `String` and perform string concatenation with result being `x = "1.02.0"`.

Similarly, the equality/inequality operators can be applied to any primitive type.

The order of preference is:

1. `Int`, `Float`: Int coerces to Float
2. `X`, `Y`: For any primitive types `X` != `Y`, both are coerced to String
3. If applying `+` to two values of the same type that cannot otherwise be summed/concatenated (i.e. Boolean, File, Directory), both values are first coerced to String

**Comment by illusional** [comment]
Yeah I like this a lot as an automatic coercion that an engine might apply. But as a pipeline author I'd like a way to explicitly coerce between types. Like force a string to be parsed as an integer or vice versa. There are a few ways that other programming languages do this:

- Provide a function (eg: `Integer.parse("1")` or `parse_integer("1")`)
- Call the type (eg: `Integer("1")`) <- this feels the most WDL to me
- Coerce using some semantic (`Int my_int = (Integer) "1'` or `"1" as Integer)

**Comment by jdidion** [comment]
Please see #373. These are separate but complementary proposals.

**Comment by jdidion** [comment]
Included in v1.1 and errata.
